contributors/design-proposals/flakiness-sla.md

```diff
@@ -0,0 +1,92 @@
+
+A pre-submit job's test result is considered to
+be a flake according to two criteria:
+
+1) it both fails and passes without any changes in the code
+or environment being tested.
+
+2) the PR in question doesn't cause the flake itself.
+
+### Measuring flakiness
+There are a number of challenges in monitoring flakiness. We expect that
+the metric will be heuristic in nature and we will iterate on it over time.
+Identifying all of the potential problems and ways to measure the metric
+are out of the scope of the document, but some currently known challenges
+are listed at the end of the document.
+
+## Flakiness SLA
+We will measure flakiness based on pull requests that are run through pre-submit
+PR testing. The metric that we will monitor is: Pre-submit flakes per PR. This metric
+will be calculated on a daily and weekly basis using:
+
+Sum(Flakes not caused by the PR) / Total(PRs)
+
+Our current SLA is that this metric will be less than 0.01
+(1% of all PRs have flakes that are not caused by the PR itself).
+
+## Activities on SLA violation
+When the project is out of SLA for flakiness on either the daily or weekly metric
+we will stop the merge queue.
+
+Only PRs with the label `queue/fix` or `queue/critical-fix` will be allowed to merge.
+The merge queue will remain blocked until a sufficient number of flake-fixes have been
+submitted to restore the daily CLA. We will measure this each day by via automation that
+re-tests a number of PRs and looking for more flakes.
+
+### Exceptions
+We may consider merging PRs under exceptional circumstances, for example
+merging security fixes or other critical bug fixes that need to be cut into
+a pre-existing release. Each exception will be considered on a case by case
+basis. Exceptions will only be granted for PRs that are already labelled with
+cherry-pick approve.
+
+## Non-goals: Flakiness at HEAD
+We will consider flakiness for PRs only, we will _not_ currently 
+measure flakiness for continuous e2e that are run at HEAD independent of PRs.
+
+There are a number of reasons for this:
+   * We already block the submit-queue if certain HEAD e2e tests are failing
+   * The rate at which code is merged at HEAD is so great that it's rare that a test will be re-run for the exact same git commit.
+   * The volume of testing on PRs is significantly higher than at HEAD, so flakes are more readily apparent.
+   * The flakiness of these tests is already being measured in other places.
+
+## Approaches to measuring the flakiness metric
+
+Currently, measuring this metric is somewhat tricky, since determining flakes
+caused by PRs vs. existing flakes is somewhat challenging. We will use a variety
+of heuristics to determine this, including looking at PRs that contain changes
+that could not cause flakes (e.g. typos, docs). As well as looking at the
+past history of test failures (e.g. if a test fails across many different PRs
+it's likely a flake)
```

Comment:
I was not having a good math day when I wrote that, but this comment isn't quite as wrong as the other one :)

You're trying to target a rate *per PR* of 1/100, but tests are run in N suites for M pushes, so the rate *per test invocation* must be 1/(100*N*M), which is probably 1/1000 or less (N is at least 5 and M is at least 2). However, this is slightly offset by the fact that we get N*M per-test samples for each PR, I guess.

I played around with a population size/sample size/confidence interval calculator, I now think we actually will get enough test runs to have a decent estimate in practice.